modules/entity_hierarchy_workbench_access/src/Plugin/AccessControlHierarchy/EntityHierarchy.php

```diff
@@ -134,130 +102,95 @@ public function defaultConfiguration() {
+      // Remove the label/id fields.
+      unset($item[$entity_type->getKey('label')], $item[$entity_type->getKey('id')]);
+      // Return just the boolean fields that were checked.
+      return array_keys(array_filter($item));
+    }, $details));
+    $tags = $entityStorage->getEntityType()->getListCacheTags();
+    foreach ($valid_ids as $id) {
+      $entity = $entityStorage->load($id);
+
+      $tags = array_merge($tags, $entity->getCacheTags());
+      foreach ($boolean_parents[$id] as $parent) {
+        $cid = sprintf('%s:%s', self::CACHE_ID, $id);
+        if ($entry = $this->cacheBackend->get($cid)) {
+          $tree[$parent][$id] = $entry->data;
+          // Cache hit.
           continue;
         }
-        $tags = array_merge($tags, $entities[$id]->getCacheTags());
-        foreach ($boolean_parents[$id] as $parent) {
-          $cid = sprintf('%s:%s', self::CACHE_ID, $id);
-          if ($entry = $this->cacheBackend->get($cid)) {
-            $tree[$parent][$id] = $entry->data;
-            // Cache hit.
-            continue;
-          }
-          $entry_tags = [];
-          $entry = [
-            'label' => $this->generateEntityLabelWithAncestry($entity, $tree_storage, $entity_type_id, $entry_tags),
-            'depth' => $node->getDepth(),
-            'parents' => [],
-            'weight' => $node->getLeft(),
-            'description' => '',
-          ];
-          $tree[$parent][$id] = $entry;
-          $this->cacheBackend->set($cid, $entry, Cache::PERMANENT, $entry_tags);
-          $tags = array_merge($tags, $entry_tags);
-        }
+        $entry_tags = [];
+        $entry = [
+          'label' => $this->generateEntityLabelWithAncestry($entity, $query_builder, $entry_tags),
+          'depth' => $query_builder->findDepth($entity),
+          'parents' => [],
+          'weight' => 1,
+          'description' => '',
+        ];
+        $tree[$parent][$id] = $entry;
+        $this->cacheBackend->set($cid, $entry, Cache::PERMANENT, $entry_tags);
+        $tags = array_merge($tags, $entry_tags);
       }
-      foreach (array_keys($tree) as $parent) {
-        uasort($tree[$parent], function (array $a, array $b) {
-          // @todo Replace this with null coalesce and spaceship operator when
-          // we only support PHP 7.0 or greater.
-          $a_weight = 0;
-          $b_weight = 0;
-          if (isset($a['weight'])) {
-            $a_weight = $a['weight'];
-          }
-          if (isset($b['weight'])) {
-            $b_weight = $b['weight'];
-          }
-          if ($a_weight < $b_weight) {
-            return -1;
-          }
-          if ($a_weight > $b_weight) {
-            return 1;
-          }
-          return 0;
-        });
-      }
-      $this->tree = $tree;
-      $this->cacheBackend->set(self::CACHE_ID, $tree, Cache::PERMANENT, array_unique($tags));
+      $entityStorage->resetCache([$id]);
```

Comment:
This tree can be huge and every row does a entity load. It's inside a loop getting the entity and processing it (so we've got past the OOM for a loadMultiple) but it still slowly builds up the entity storage cache. The `resetCache()` clears that out (consider it garbage collection).

Without this change, the access control hierarchy management form crashes with an OOM quite quickly. Doesn't crash on DCS with change (just slow).

Partly caused by the poor filtering of objects (it doesn't filter available entity bundle in the queries, it just let you use them). Not something I wanted to take on the refactor for because we don't have it in use.

Reply:
Can we get a comment to that effect - makes sense